Removes unnecessary file util

CMurtagh-LGTM opened the pull request:
As c++17 has all these utils in the standard, we can delete our version.
Resolves #484

CMurtagh-LGTM committed:
removed old file util

Bidski commented:
Nooooo! now I have to rewrite this stuff in c++17 :'(

CMurtagh-LGTM commented:
Read https://en.cppreference.com/w/cpp/filesystem and https://en.cppreference.com/w/cpp/io.

Bidski commented:
Also, remove your changes to `.vscode/settings.json`

Bidski commented:
> Read https://en.cppreference.com/w/cpp/filesystem and https://en.cppreference.com/w/cpp/io.

I know, but effort :P

CMurtagh-LGTM commented:
> > Read https://en.cppreference.com/w/cpp/filesystem and https://en.cppreference.com/w/cpp/io.
> 
> I know, but effort :P

Kip was planning on Ben to fix all that for us XD.

CMurtagh-LGTM committed:
Remove .vscode changes

Bidski commented:
> > > Read https://en.cppreference.com/w/cpp/filesystem and https://en.cppreference.com/w/cpp/io.
> > 
> > 
> > I know, but effort :P
> 
> Kip was planning on Ben to fix all that for us.

I assume another part of this PR is going to fix the compile errors?

KipHamiltons commented:
@xXBenjaminYoungXx were you still interested in this one? It's fine if not, and there's not a real hurry

KipHamiltons commented:
> Reading through this PR and the docs Ysobel linked earlier, it seems like shared/utility/file/fileutil.cpp has some functions that don't exist in the standard libraries?
> 
> e.g.
> 
> * getModificationTime
> * listDir
> * pathSplit
> * touch
>

`std::filesystem` is fairly complete. It has pretty much all of these:
We can get modification time with [last_write_time](https://en.cppreference.com/w/cpp/filesystem/last_write_time). `listDir` can be emulated by constructor a directory iterator with the path [(relevant SO)](https://stackoverflow.com/a/612176/11108223) (although we don't actually use it anyway :shrug: ). `pathSplit` is doing what `std::path`'s `stem` and `extension` do, with some other stuff. `touch` is something that isn't in `std::filesystem`, but we don't use it and that it's reasonably trivial to create a file with a `std::ostream`, so I don't think that'll be a huge issue.
 
> (Some of these are never used yet which could be a fair reason to remove them from master (now main), but leaving their implementation in fileutil so it is used consistently throughout other files is certainly much cleaner from a design cohesion PoV if they are used in future)
> 
> Not sure whether I'm right on this, it just seems a little wasteful to remove fileutil and then spend time edit the existing files (ScriptTuner, ScriptEngine, Configuration, and Script from what I could find in this branch, it may be more in others) to use standard libs and then editing them all until it builds again, without an advantage in doing so other than cleaning up the file structure?

There are a few big benefits for using standard libraries (for all things - not limiting this to this example) when possible IMO:
- They're optimised/optimal for most things. You're not going to get better perf that the standard library unless you have specific information/assumptions you can make. For example, `std::sort` has an optimal time complexity algorithm (faster than `n log n` best case, `n log n` worst case AFAIK), although I am sure you could beat it if you were targeting a specific CPU with known cache sizes or you can make assumptions about how sorted the list is.
- You don't have to maintain them. They'll (almost always) do _exactly_ what they're meant to. They'll pass their tests because huge teams of people make sure they're good to go. You never have to update them. No formatting or clang-tidy changes. Nothing. This is huge.
- They're clearly documented, with tutorials and stack-overflow answers and heaaaps of stuff online to teach you how and when to use a thing. (we also don't have to maintain their docs).
- They're portable. All platforms (with few exceptions), without extra effort. They encapsulate the hard parts, providing a clean interface for all of the platforms. This isn't relevant for us because we are running on docker, but this is just a general thing.

Keeping code around because it could be useful later is generally bad. It's still in version control - it's not gone forever. However, it clutters things, takes up maintenance time, and just generally rots. It's either being maintained up to the standard of the code around it (bad, because you're wasting time on code that you don't use), or it's not being maintained, and it's rotting beyond repair (also bad, because it's unused and unuseable).

This response ending up being a little verbose. Hopefully it comes across ok :sweat_smile: Long story short, I am keen to remove this in favour of `std::filesystem` :stuck_out_tongue:

BryceTuppurainen committed:
Cutting out exists

Added the deleted files back to test that the code builds correctly, added todos to identify where fileutil needs to be removed, and replaced utility::file::exists with std::filesystem::exists where applicable across files I found that used it.

TrentHouliston reviewed (commented):
Lots of the places where you're using exists now should be is_regular_file

andrewdeanphillips committed:
Removed loadFromFile, getModificationTime, makeDir, and listFiles functions from fileutil

andrewdeanphillips committed:
Trigger Build

andrewdeanphillips committed:
Removed isDir, listDir, touch

BryceTuppurainen committed:
Corrected most locations to is_regular_file

There were two exceptions to the rule where `exists();` seems to be more appropriate than `is_regular_file();`

BryceTuppurainen commented on module/behaviour/tools/ScriptTuner/src/ScriptTuner.cpp:
```diff
@@ -73,7 +74,7 @@ namespace module::behaviour::tools {
 
                 // Check if the script exists and load it if it does.
-                if (utility::file::exists(scriptPath)) {
+                if (std::filesystem::exists(scriptPath)) {
```
Thanks for letting me know. I've updated this in the most recent commit in most places.

BryceTuppurainen committed:
Removed isDir

isDir was completely unused in our codebase, I've removed it from both the fileutil files.

BryceTuppurainen committed:
Tidied up the HPP preparing to remove

BryceTuppurainen committed:
Replaced the single implentation of pathSplit

Completely purged the file folder in tools\Firmware

andrewdeanphillips committed:
Merge to main

andrewdeanphillips committed:
Replaced writeToFile function used in ScriptTuner

andrewdeanphillips committed:
deleted fileutil

andrewdeanphillips committed:
Included fstream in Script.hpp

andrewdeanphillips committed:
Included fstream in Configuration.hpp

andrewdeanphillips committed:
Added pathSplit function from deleted fileutil into ScriptEngine

andrewdeanphillips committed:
Updated call to the pathsplit function in ScriptEngine

andrewdeanphillips committed:
Correct pathSplit function defintion

andrewdeanphillips committed:
Added readFile from fileutil as a local function of FakeCamera

andrewdeanphillips committed:
Add include filesystem to FakeCamera.hpp

andrewdeanphillips committed:
Merge with main

thomaslegge98 committed:
Updated settings.json with main version

Exactly the same as this current PRs settings.json, but with different tab spacing

ysims committed:
Merge branch 'main' into cmurtagh/filesystem

ysims committed:
format

ysims committed:
Add directives and make snake case
